File: app/android/app/src/main/kotlin/com/uae4arm2026/AmigaTexturePlugin.kt
package com.uae4arm2026

import android.view.Choreographer
import android.view.Surface
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.plugin.common.MethodChannel
import io.flutter.view.TextureRegistry

/**
 * The Amiga's picture as an external texture.
 *
 * The panel's first version took each frame the long way round: a native
 * staging copy, a Dart allocation, decodeImageFromPixels, and a fresh GPU
 * upload -- four copies of 1.7MB per frame at 752x576, which is why the
 * Android panel had to be capped at 30fps to leave the audio callback any
 * room. Flutter can composite a surface the app owns instead, and that turns
 * the whole thing into one memcpy from the emulator's front buffer into the
 * buffer the compositor is about to read.
 *
 * ## Why a SurfaceTexture and not a SurfaceProducer
 *
 * The obvious registry entry is `createSurfaceProducer()`, and it is the one
 * this used at first. It does not work for a CPU producer. On API 29+ the
 * producer is an ImageReader built with `ImageFormat.PRIVATE` and
 * `HardwareBuffer.USAGE_GPU_SAMPLED_IMAGE` -- an opaque, GPU-only buffer.
 * `ANativeWindow_lock` on it has nothing CPU-writable to hand back, so it
 * fails; every present failed with it, and because failing to present is not
 * an error anywhere in the path, the result was a black panel with working
 * sound. `setSize()` on it does not help: the format is the problem, not the
 * geometry.
 *
 * A SurfaceTexture's surface is a plain BufferQueue. It hands out CPU-writable
 * buffers, it takes its geometry from the producer -- so the native side can
 * follow a mode change immediately rather than waiting to be resized -- and
 * posting a buffer marks the texture frame available without anything having
 * to call `scheduleFrame`.
 *
 * Both ends still have to agree on the size, though: the producer's geometry
 * is set natively and the consumer's default buffer size is set here, in
 * [resizeToFrame], from the same number. Letting them drift is what draws the
 * picture into the corner of the panel.
 *
 * Nothing about a frame passes through Dart or through a platform channel.
 * The channel is used twice a session -- to hand the texture id up and to tear
 * it down -- and the frames themselves are pushed by a Choreographer callback
 * calling straight into the native side.
 */
class AmigaTexturePlugin(private val registry: TextureRegistry) {

	companion object {
		const val CHANNEL = "uae4arm2026/texture"

		/**
		 * The library is the emulator's own, and it is tens of megabytes.
		 * Loading it here would slow every cold start for a screen most
		 * sessions reach only after choosing something to run; by the time a
		 * texture is asked for, the in-process core has loaded it already.
		 */
		@JvmStatic external fun nativeAttachSurface(surface: android.view.Surface?): Boolean
		@JvmStatic external fun nativeDetachSurface()
		@JvmStatic external fun nativePresent(): Boolean
		@JvmStatic external fun nativeFrameSize(): Long
	}

	private var entry: TextureRegistry.SurfaceTextureEntry? = null
	private var surface: Surface? = null
	private var frameCallback: Choreographer.FrameCallback? = null

	/** What the SurfaceTexture is currently sized to. */
	private var width = 0
	private var height = 0

	fun register(engine: FlutterEngine) {
		MethodChannel(engine.dartExecutor.binaryMessenger, CHANNEL)
			.setMethodCallHandler { call, result ->
				when (call.method) {
					"create" -> result.success(create())
					"dispose" -> {
						dispose()
						result.success(true)
					}
					else -> result.notImplemented()
				}
			}
	}

	private fun create(): Map<String, Any>? {
		dispose()
		val created = try {
			registry.createSurfaceTexture()
		} catch (error: Throwable) {
			// No entry means no texture path, which is a fallback and not a
			// failure: Dart takes the copy-and-decode route instead.
			return null
		}
		entry = created
		// A first size so there is a buffer queue before the Amiga has drawn
		// anything. Both ends move off it together -- native through
		// ANativeWindow_setBuffersGeometry, this side through resizeToFrame --
		// the moment a frame reports a different mode.
		width = 720
		height = 568
		created.surfaceTexture().setDefaultBufferSize(width, height)
		val target = Surface(created.surfaceTexture())
		surface = target
		if (!attach(target)) {
			dispose()
			return null
		}
		startFrameLoop()
		return mapOf("id" to created.id(), "width" to width, "height" to height)
	}

	private fun attach(target: Surface): Boolean =
		try {
			nativeAttachSurface(target)
		} catch (error: UnsatisfiedLinkError) {
			// A core built before host_texture.cpp existed. Same answer as no
			// entry at all: fall back.
			false
		}

	private fun startFrameLoop() {
		val choreographer = Choreographer.getInstance()
		val callback = object : Choreographer.FrameCallback {
			override fun doFrame(frameTimeNanos: Long) {
				if (entry == null) return
				resizeToFrame()
				// Posting a buffer is what marks the texture frame available,
				// so there is nothing to schedule.
				nativePresent()
				choreographer.postFrameCallback(this)
			}
		}
		frameCallback = callback
		choreographer.postFrameCallback(callback)
	}

	/**
	 * Keeps the CONSUMER's idea of the buffer size in step with the producer's.
	 *
	 * The native side sets the producer geometry itself, which is what lets a
	 * mode change take effect immediately. That is only half of it: a
	 * SurfaceTexture also has a default buffer size, and Flutter samples the
	 * texture through the transform matrix that goes with it. Let the two drift
	 * apart -- native producing 752x576, or the 1920x1080 the headless surface
	 * fallback can publish, into a texture still declared 720x568 -- and the
	 * compositor draws the frame at the wrong scale, which on screen is the
	 * picture shrunk into the top-left corner of the panel.
	 *
	 * That is the "resuming a game shrinks it to the top left" report. Resume
	 * is where it shows up first because a restored state brings its own
	 * display mode, which is usually not the one the surface was created with.
	 */
	private fun resizeToFrame() {
		val created = entry ?: return
		val packed = nativeFrameSize()
		val frameWidth = (packed ushr 32).toInt()
		val frameHeight = (packed and 0xFFFFFFFFL).toInt()
		if (frameWidth <= 0 || frameHeight <= 0) return
		if (frameWidth == width && frameHeight == height) return
		width = frameWidth
		height = frameHeight
		created.surfaceTexture().setDefaultBufferSize(width, height)
	}

	/** Releases the surface and stops the frame loop. Safe to call twice. */
	fun dispose() {
		frameCallback?.let { Choreographer.getInstance().removeFrameCallback(it) }
		frameCallback = null
		try {
			// Sink first: present() must stop before the surface it writes to
			// is released, not after.
			nativeDetachSurface()
		} catch (error: UnsatisfiedLinkError) {
			// Never attached; nothing to let go of.
		}
		surface?.release()
		surface = null
		entry?.release()
		entry = null
		width = 0
		height = 0
	}
}

File: app/ios/Runner/AmigaTexturePlugin.swift
import Accelerate
import CoreVideo
import Flutter
import QuartzCore

/// The Amiga's picture as a Flutter external texture.
///
/// The panel's first version took every frame the long way: a native staging
/// copy, a Dart allocation, `decodeImageFromPixels`, and a fresh GPU upload.
/// At 752x576 that is close to 7MB of copying per frame, and it is why the
/// panel had to be capped below the Amiga's own frame rate. A `FlutterTexture`
/// hands the compositor a buffer the app owns, so a frame costs one strided
/// copy out of the emulator's front buffer and a channel permute.
///
/// Frames are pushed by a `CADisplayLink` rather than polled from Dart, so
/// nothing about a frame crosses a platform channel. The channel is used
/// twice a session: once to hand the texture id up, once to tear it down.
///
/// Everything here is best-effort. A core too old to carry host_texture.cpp
/// has no symbols to bind, `create` answers nil, and the Dart side falls back
/// to copy-and-decode with nothing to configure.
final class AmigaTexturePlugin: NSObject {

  static let channelName = "uae4arm2026/texture"

  private typealias SinkFn = @convention(c) (UnsafeMutableRawPointer?) -> Bool
  private typealias SetSinkFn =
    @convention(c) (SinkFn?, UnsafeMutableRawPointer?) -> Void
  private typealias PresentFn = @convention(c) () -> Bool
  private typealias SizeFn =
    @convention(c) (UnsafeMutablePointer<Int32>?, UnsafeMutablePointer<Int32>?) -> Void
  private typealias CopyFn = @convention(c) (
    UnsafeMutableRawPointer?, Int32, Int32,
    UnsafeMutablePointer<Int32>?, UnsafeMutablePointer<Int32>?,
    UnsafeMutablePointer<UInt64>?
  ) -> Int32

  private let registry: FlutterTextureRegistry
  private var textureId: Int64?
  private var displayLink: CADisplayLink?

  /// Guards [ready] against the raster thread, which asks for it through
  /// copyPixelBuffer while the display link may be filling the next one.
  private let lock = NSLock()

  /// The buffer the compositor should draw, and the one being written. Two,
  /// not one: writing into the buffer the raster thread is reading is a tear
  /// that only shows on slow frames.
  private var ready: CVPixelBuffer?
  private var spare: CVPixelBuffer?
  private var bufferWidth: Int32 = 0
  private var bufferHeight: Int32 = 0

  private var setSink: SetSinkFn?
  private var present: PresentFn?
  private var frameSize: SizeFn?
  private var copyStrided: CopyFn?

  init(registry: FlutterTextureRegistry) {
    self.registry = registry
    super.init()
  }

  static func register(with controller: FlutterViewController) -> AmigaTexturePlugin? {
    guard
      let registrar = controller.engine?.registrar(forPlugin: "AmigaTexturePlugin"),
      let textures = registrar.textures() as FlutterTextureRegistry?
    else { return nil }

    let plugin = AmigaTexturePlugin(registry: textures)
    let channel = FlutterMethodChannel(
      name: channelName, binaryMessenger: controller.binaryMessenger)
    channel.setMethodCallHandler { call, result in
      switch call.method {
      case "create":
        result(plugin.create())
      case "dispose":
        plugin.dispose()
        result(true)
      default:
        result(FlutterMethodNotImplemented)
      }
    }
    return plugin
  }

  // MARK: - Lifecycle

  private func bindSymbols() -> Bool {
    if setSink != nil { return true }
    guard
      let sinkSymbol = EmulatorHost.shared.coreSymbol("uae4arm_host_texture_set_sink"),
      let presentSymbol = EmulatorHost.shared.coreSymbol("uae4arm_host_texture_present"),
      let sizeSymbol = EmulatorHost.shared.coreSymbol("uae4arm_host_framebuffer_size"),
      let copySymbol = EmulatorHost.shared.coreSymbol("uae4arm_host_copy_framebuffer_strided")
    else { return false }
    setSink = unsafeBitCast(sinkSymbol, to: SetSinkFn.self)
    present = unsafeBitCast(presentSymbol, to: PresentFn.self)
    frameSize = unsafeBitCast(sizeSymbol, to: SizeFn.self)
    copyStrided = unsafeBitCast(copySymbol, to: CopyFn.self)
    return true
  }

  private func create() -> [String: Any]? {
    dispose()
    guard bindSymbols() else { return nil }

    // A first size so there is a texture before the Amiga has drawn anything.
    // Replaced the moment a real frame reports its mode.
    guard resize(width: 720, height: 568) else { return nil }

    let id = registry.registerTexture(self)
    textureId = id

    setSink?(
      { context in
        guard let context = context else { return false }
        let plugin = Unmanaged<AmigaTexturePlugin>
          .fromOpaque(context).takeUnretainedValue()
        return plugin.fill()
      }, Unmanaged.passUnretained(self).toOpaque())

    let link = CADisplayLink(target: self, selector: #selector(step))
    link.add(to: .main, forMode: .common)
    displayLink = link

    return ["id": id, "width": Int(bufferWidth), "height": Int(bufferHeight)]
  }

  func dispose() {
    displayLink?.invalidate()
    displayLink = nil
    // Sink first: it must stop before the buffers it writes into are released.
    setSink?(nil, nil)
    if let id = textureId {
      registry.unregisterTexture(id)
      textureId = nil
    }
    lock.lock()
    ready = nil
    spare = nil
    bufferWidth = 0
    bufferHeight = 0
    lock.unlock()
  }

  // MARK: - Frames

  @objc private func step() {
    guard let present = present, let frameSize = frameSize else { return }

    // The Amiga changes display mode mid-game, and fill() refuses a buffer
    // that is not exactly the frame's size rather than shear the picture or
    // leave a margin of stale pixels. Resizing is what lets it start again.
    var width: Int32 = 0
    var height: Int32 = 0
    frameSize(&width, &height)
    if width > 0, height > 0, width != bufferWidth || height != bufferHeight {
      _ = resize(width: width, height: height)
      return
    }

    if present(), let id = textureId {
      registry.textureFrameAvailable(id)
    }
  }

  /// Called from the core's present, with no emulator lock held.
  private func fill() -> Bool {
    guard let copyStrided = copyStrided else { return false }

    lock.lock()
    let target = spare
    let width = bufferWidth
    let height = bufferHeight
    lock.unlock()

    guard let buffer = target, width > 0, height > 0 else { return false }

    CVPixelBufferLockBaseAddress(buffer, [])
    defer { CVPixelBufferUnlockBaseAddress(buffer, []) }
    guard let base = CVPixelBufferGetBaseAddress(buffer) else { return false }

    let stride = CVPixelBufferGetBytesPerRow(buffer)
    let written = copyStrided(
      base, Int32(stride / 4), height, nil, nil, nil)
    if written <= 0 { return false }

    // The emulator publishes ABGR8888, which little-endian is R,G,B,A in
    // memory; a Flutter texture wants BGRA. Doing it here rather than asking
    // the core for another format keeps Android -- where R,G,B,A is exactly
    // what the compositor wants -- on a copy with no fixup at all. vImage
    // rather than a loop: this is a SIMD permute over 1.7MB, and a scalar
    // version of it costs more than the copy it follows.
    var image = vImage_Buffer(
      data: base, height: vImagePixelCount(height),
      width: vImagePixelCount(width), rowBytes: stride)
    let toBGRA: [UInt8] = [2, 1, 0, 3]
    _ = vImagePermuteChannels_ARGB8888(&image, &image, toBGRA, vImage_Flags(kvImageNoFlags))

    // ...and the alpha byte forced opaque, HERE.
    //
    // The emulator leaves the top byte zero, and this buffer is 32BGRA, which
    // Core Video and Flutter both take at its word: every pixel of a
    // perfectly correct picture is fully transparent and the screen shows
    // whatever is behind it.
    //
    // It used to be done by the publisher, for every frame on every platform.
    // That is two million read-modify-writes a frame at the resolutions an
    // RTG collection asks for, on the thread whose scheduling latency causes
    // audio underruns -- so it moved to the consumers that actually care.
    // Android does not: its window is RGBX, where the byte is ignored by
    // definition. iOS does, and pays for it in one more SIMD pass over a
    // buffer already in cache from the permute above.
    _ = vImageOverwriteChannelsWithScalar_ARGB8888(
      255, &image, &image, 0x1, vImage_Flags(kvImageNoFlags))

    lock.lock()
    spare = ready
    ready = buffer
    lock.unlock()
    return true
  }

  /// Allocates the pair of buffers for a new mode. Returns false if the
  /// allocation failed, in which case the texture stays on its old size and
  /// the next display link tick tries again.
  private func resize(width: Int32, height: Int32) -> Bool {
    guard let first = makeBuffer(width: width, height: height),
      let second = makeBuffer(width: width, height: height)
    else { return false }
    lock.lock()
    ready = first
    spare = second
    bufferWidth = width
    bufferHeight = height
    lock.unlock()
    return true
  }

  private func makeBuffer(width: Int32, height: Int32) -> CVPixelBuffer? {
    var buffer: CVPixelBuffer?
    let attributes: [CFString: Any] = [
      // Both are needed: Metal to sample it, IOSurface because a Flutter
      // texture is handed to the compositor rather than drawn here.
      kCVPixelBufferMetalCompatibilityKey: true,
      kCVPixelBufferIOSurfacePropertiesKey: [:] as CFDictionary,
    ]
    let status = CVPixelBufferCreate(
      kCFAllocatorDefault, Int(width), Int(height),
      kCVPixelFormatType_32BGRA, attributes as CFDictionary, &buffer)
    return status == kCVReturnSuccess ? buffer : nil
  }
}

extension AmigaTexturePlugin: FlutterTexture {
  func copyPixelBuffer() -> Unmanaged<CVPixelBuffer>? {
    lock.lock()
    defer { lock.unlock() }
    guard let buffer = ready else { return nil }
    return Unmanaged.passRetained(buffer)
  }
}
